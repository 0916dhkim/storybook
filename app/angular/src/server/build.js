// Building for distribution defaults to production target.
process.env.NODE_ENV = process.env.NODE_ENV || 'production';

// Note: We are importing via CommonJS to be able to adjust NODE_ENV
// to production first before proceeding. This is required to correctly
// e.g. selecting the `.env` files to import. It also helps with all ESM
// imports which as some side-effect decide on things at "loadtime".
const server = require('@storybook/core/server');
server.buildStatic(require('./options').default);
