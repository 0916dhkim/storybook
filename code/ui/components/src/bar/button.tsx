import type {
  AnchorHTMLAttributes,
  ButtonHTMLAttributes,
  DetailedHTMLProps,
  ForwardedRef,
  ForwardRefExoticComponent,
  ReactElement,
  RefAttributes,
} from 'react';
import React, { forwardRef } from 'react';
import { styled, isPropValid } from '@storybook/theming';
import { transparentize } from 'polished';
import { auto } from '@popperjs/core';

interface ButtonProps
  extends DetailedHTMLProps<ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement> {
  href?: never;
  target?: never;
}
interface LinkProps
  extends DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement> {
  disabled?: void;
  href: string;
}

const isLink = (obj: {
  props: ButtonProps | LinkProps;
  ref: ForwardedRef<HTMLButtonElement> | ForwardedRef<HTMLAnchorElement>;
}): obj is { props: LinkProps; ref: ForwardedRef<HTMLAnchorElement> } => {
  return typeof obj.props.href === 'string';
};

const isButton = (obj: {
  props: ButtonProps | LinkProps;
  ref: ForwardedRef<HTMLButtonElement> | ForwardedRef<HTMLAnchorElement>;
}): obj is { props: ButtonProps; ref: ForwardedRef<HTMLButtonElement> } => {
  return typeof obj.props.href !== 'string';
};

function ForwardRefFunction(p: LinkProps, r: ForwardedRef<HTMLAnchorElement>): ReactElement;
function ForwardRefFunction(p: ButtonProps, r: ForwardedRef<HTMLButtonElement>): ReactElement;
function ForwardRefFunction(
  { children, ...rest }: ButtonProps | LinkProps,
  ref: ForwardedRef<HTMLButtonElement> | ForwardedRef<HTMLAnchorElement>
) {
  const o = { props: rest, ref };
  if (isLink(o)) {
    return (
      <a ref={o.ref} {...o.props}>
        {children}
      </a>
    );
  }
  if (isButton(o)) {
    return (
      <button ref={o.ref} type="button" {...o.props}>
        {children}
      </button>
    );
  }
  throw new Error('invalid props');
}
type ButtonLike<P = {}> = ForwardRefExoticComponent<
  Omit<ButtonProps, 'ref'> & RefAttributes<HTMLButtonElement> & P
>;

type LinkLike<P = {}> = ForwardRefExoticComponent<
  Omit<LinkProps, 'ref'> & RefAttributes<HTMLAnchorElement> & P
>;

const ButtonOrLink: ButtonLike | LinkLike = forwardRef(ForwardRefFunction) as ButtonLike | LinkLike;

ButtonOrLink.displayName = 'ButtonOrLink';

export interface TabButtonProps {
  active?: boolean;
  textColor?: string;
}

export const TabButton = styled(ButtonOrLink, { shouldForwardProp: isPropValid })<TabButtonProps>(
  {
    whiteSpace: 'normal',
    display: 'inline-flex',
    overflow: 'hidden',
    verticalAlign: 'top',
    justifyContent: 'center',
    alignItems: 'center',
    textAlign: 'center',
    textDecoration: 'none',

    '&:empty': {
      display: 'none',
    },
  },
  ({ theme }) => ({
    padding: '0 15px',
    transition: 'color 0.2s linear, border-bottom-color 0.2s linear',
    height: 40,
    lineHeight: '12px',
    cursor: 'pointer',
    background: 'transparent',
    border: '0 solid transparent',
    borderTop: '3px solid transparent',
    borderBottom: '3px solid transparent',
    fontWeight: 'bold',
    fontSize: 13,

    '&:focus': {
      outline: '0 none',
      borderBottomColor: theme.color.secondary,
    },
  }),
  ({ active, textColor, theme }) =>
    active
      ? {
          color: textColor || theme.barSelectedColor,
          borderBottomColor: theme.barSelectedColor,
        }
      : {
          color: textColor || theme.barTextColor,
          borderBottomColor: 'transparent',
        }
);
TabButton.displayName = 'TabButton';

export interface IconButtonProps {
  active?: boolean;
  disabled?: boolean;
}

export const IconButton = styled(ButtonOrLink, { shouldForwardProp: isPropValid })<IconButtonProps>(
  () => ({
    alignItems: 'center',
    background: 'transparent',
    border: 'none',
    borderRadius: 4,
    color: 'inherit',
    cursor: 'pointer',
    display: 'inline-flex',
    fontSize: 13,
    fontWeight: 'bold',
    height: 28,
    justifyContent: 'center',
    marginTop: 6,
    padding: '8px 7px',

    '& > svg': {
      width: 14,
    },
  }),
  ({ active, theme }) =>
    active
      ? {
          backgroundColor: theme.background.hoverable,
          color: theme.color.secondary,
        }
      : {},
  ({ disabled, theme }) =>
    disabled
      ? {
          opacity: 0.5,
          cursor: 'not-allowed',
        }
      : {
          '&:hover, &:focus-visible': {
            background: transparentize(0.88, theme.color.secondary),
            color: theme.color.secondary,
          },
          '&:focus-visible': {
            outline: auto, // Ensures links have the same focus style
          },
          '&:focus:not(:focus-visible)': {
            outline: 'none',
          },
        }
);
IconButton.displayName = 'IconButton';

const IconPlaceholder = styled.div(({ theme }) => ({
  width: 14,
  height: 14,
  backgroundColor: theme.appBorderColor,
  animation: `${theme.animation.glow} 1.5s ease-in-out infinite`,
}));

const IconButtonSkeletonWrapper = styled.div(() => ({
  marginTop: 6,
  padding: 7,
  height: 28,
}));

export const IconButtonSkeleton = () => (
  <IconButtonSkeletonWrapper>
    <IconPlaceholder />
  </IconButtonSkeletonWrapper>
);
