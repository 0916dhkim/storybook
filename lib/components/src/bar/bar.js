import { styled } from '@storybook/theming';

export const Bar = styled.div(
  ({ theme }) => ({
    display: 'flex',
    justifyContent: 'space-between',
    background: theme.barFill,
    color: theme.barTextColor,
    overflow: 'auto',
    height: 40,
    position: 'relative',
  }),
  ({ theme, border }) =>
    border
      ? {
          background: `${theme.barFill} linear-gradient(to bottom, transparent calc(100% - 1px), ${
            theme.mainBorderColor
          } calc(100% - 1px))`,
        }
      : {}
);
