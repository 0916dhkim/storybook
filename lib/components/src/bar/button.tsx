import React, { AnchorHTMLAttributes, ButtonHTMLAttributes, DetailedHTMLProps } from 'react';
import { styled, isPropValid } from '@storybook/theming';
import { transparentize } from 'polished';

interface ButtonProps
  extends DetailedHTMLProps<ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement> {
  href?: void;
}
interface LinkProps
  extends DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement> {
  href: string;
}

const ButtonOrLink = ({ children, ...restProps }: ButtonProps | LinkProps) =>
  restProps.href != null ? (
    <a {...(restProps as LinkProps)}>{children}</a>
  ) : (
    <button type="button" {...(restProps as ButtonProps)}>
      {children}
    </button>
  );

export interface TabButtonProps {
  active?: boolean;
  textColor?: string;
}

export const TabButton = styled(ButtonOrLink, { shouldForwardProp: isPropValid })<TabButtonProps>(
  {
    whiteSpace: 'normal',
    display: 'inline-flex',
    overflow: 'hidden',
    verticalAlign: 'top',
    justifyContent: 'center',
    alignItems: 'center',
    textAlign: 'center',
    textDecoration: 'none',

    '&:empty': {
      display: 'none',
    },
  },
  ({ theme }) => ({
    padding: '0 15px',
    transition: 'color 0.2s linear, border-bottom-color 0.2s linear',
    height: 40,
    lineHeight: '12px',
    cursor: 'pointer',
    background: 'transparent',
    border: '0 solid transparent',
    borderTop: '3px solid transparent',
    borderBottom: '3px solid transparent',
    fontWeight: 'bold',
    fontSize: 13,

    '&:focus': {
      outline: '0 none',
      borderBottomColor: theme.color.secondary,
    },
  }),
  ({ active, textColor, theme }) =>
    active
      ? {
          color: textColor || theme.barSelectedColor,
          borderBottomColor: theme.barSelectedColor,
        }
      : {
          color: textColor || theme.barTextColor,
          borderBottomColor: 'transparent',
        }
);
TabButton.displayName = 'TabButton';

export interface IconButtonProps {
  active?: boolean;
}

export const IconButton = styled(ButtonOrLink, { shouldForwardProp: isPropValid })<IconButtonProps>(
  ({ theme }) => ({
    display: 'inline-flex',
    justifyContent: 'center',
    alignItems: 'center',
    height: 24,
    marginTop: 8,
    padding: '6px 5px',
    background: 'transparent',
    border: 'none',
    borderRadius: 4,
    color: 'inherit',
    cursor: 'pointer',

    // Icon Buttons may have text depending on user preferences.
    // While we don't recommend having text for IconButtons, this style ensures that the text is the correct size.
    fontWeight: 'bold',
    fontSize: 13,

    // I am unsure we need this transition, but if you want a smooth animation, this works nicely
    // transition: 'all 0.15s ease-in-out',

    '&:hover, &:focus': {
      outline: '0 none',
      background: theme.background.hoverable,
      color: theme.color.secondary,
    },
    '& > svg': {
      width: 14,
    },
  }),
  ({ active, theme }) =>
    active
      ? {
          outline: '0 none',
          backgroundColor: theme.color.secondary,
          color: theme.color.inverseText,

          '&:hover, &:focus': {
            backgroundColor: transparentize(0.1, theme.color.secondary),
            color: theme.color.inverseText,
          },
        }
      : {}
);
IconButton.displayName = 'IconButton';
