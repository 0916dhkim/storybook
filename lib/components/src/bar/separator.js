import React, { Fragment } from 'react';
import { styled } from '@storybook/theming';

export const interleaveSeparators = list =>
  list.reduce(
    (acc, item, index) =>
      item ? (
        <Fragment key={item.id || item.key || `f-${index}`}>
          {acc}
          {index > 0 ? <Separator key={`s-${index}`} /> : null}
          {item.render() || item}
        </Fragment>
      ) : (
        acc
      ),
    null
  );

export const Separator = styled.span(
  ({ theme }) => ({
    width: 1,
    height: 24,
    background: theme.mainBorderColor,
    marginTop: 8,
  }),
  {
    '& + &': {
      display: 'none',
    },
  }
);
