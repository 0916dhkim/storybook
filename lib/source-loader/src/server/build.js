import { readStory } from './dependencies-lookup/readAsObject';
import { getRidOfUselessFilePrefixes } from './dependencies-lookup/getRidOfUselessFilePrefixes';

export function insertAfterImports(insert, source) {
  let start = source.lastIndexOf('\nimport ');
  if (start === -1) {
    start = 0;
  } else {
    start = 1 + source.indexOf('\n', start + 1);
  }
  const imports = source.slice(0, start);
  const rest = source.slice(start);
  return `${imports}${insert}${rest}`;
}

export function transform(inputSource) {
  return readStory(this, inputSource)
    .then(getRidOfUselessFilePrefixes)
    .then(
      ({
        prefix,
        resource,
        source,
        sourceJson,
        addsMap,
        dependencies,
        localDependencies,
        idsToFrameworks,
      }) => {
        const preamble = `
// @ts-ignore
var withSourceLoader = require('@storybook/source-loader/preview').withSource;
// @ts-ignore
var addSourceDecorator = require("@storybook/source-loader/preview").addSource;
// @ts-ignore
var __SOURCE_PREFIX__ = "${prefix.replace(/\\([^\\ ])/g, '\\\\$1')}";
// @ts-ignore
var __STORY__ = ${sourceJson};
// @ts-ignore
var __ADDS_MAP__ = ${JSON.stringify(addsMap)};
// @ts-ignore
var __MAIN_FILE_LOCATION__ = ${JSON.stringify(resource)};
// @ts-ignore
var __MODULE_DEPENDENCIES__ = ${JSON.stringify(dependencies)};
// @ts-ignore
var __LOCAL_DEPENDENCIES__ = ${JSON.stringify(localDependencies)};
// @ts-ignore
var __IDS_TO_FRAMEWORKS__ = ${JSON.stringify(idsToFrameworks)};
        `;
        const annotated = insertAfterImports(preamble, source);
        return annotated;
      }
    );
}
