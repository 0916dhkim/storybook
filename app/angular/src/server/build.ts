import { buildStatic } from '@storybook/core/server';
import { logger } from '@storybook/node-logger';
import options from './options';

async function build() {
  try {
    await buildStatic(options);

    // #15227
    process.exit(0);
  } catch (error) {
    logger.error(error);
  }
}

build();
