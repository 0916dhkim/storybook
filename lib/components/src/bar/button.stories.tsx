import React from 'react';

import { IconButton } from './button';
import { Icons } from '../icon/icon';

export default {
  component: IconButton,
  title: 'Basics/IconButton',
};

// @TODO: Discuss about this use case
/* eslint-disable-next-line storybook/prefer-pascal-case, no-underscore-dangle */
export const _IconButton = () => (
  <IconButton>
    <Icons icon="bookmark" />
  </IconButton>
);

export const Active = () => (
  <IconButton active>
    <Icons icon="beaker" />
  </IconButton>
);

export const WithText = () => (
  <IconButton>
    <Icons icon="circlehollow" />
    &nbsp;Howdy!
  </IconButton>
);

export const WithTextActive = () => (
  <IconButton active>
    <Icons icon="circlehollow" />
    &nbsp;Howdy!
  </IconButton>
);
