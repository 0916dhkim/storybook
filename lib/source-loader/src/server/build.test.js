import { insertAfterImports } from './build';

const insert = 'INSERT\n';

describe('insertAfterImports', () => {
  it('no imports', () => {
    const noImports = `
foo bar;
baz;
    `.trim();
    expect(insertAfterImports(insert, noImports)).toMatchSnapshot();
  });

  it('imports', () => {
    const hasImports = `
import foo;
import bar;
whatever;
    `.trim();
    expect(insertAfterImports(insert, hasImports)).toMatchSnapshot();
  });
});
