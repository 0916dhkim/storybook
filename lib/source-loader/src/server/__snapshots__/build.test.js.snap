// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insertAfterImports addon-notes 1`] = `
"import React from 'react';

import BaseButton from '../components/BaseButton';
import markdownNotes from './notes/notes.md';
INSERT

import React from 'react';

import BaseButton from '../components/BaseButton';
import markdownNotes from './notes/notes.md';

const markdownString = '...';"
`;

exports[`insertAfterImports imports 1`] = `
"import './foo';
import './bar';
INSERT

import './foo';
import './bar';
whatever;"
`;

exports[`insertAfterImports multi-line imports 1`] = `
"import 'foo';
import {
  bar
} from 'baz';
INSERT

import 'foo';
import {
  bar
} from 'baz';
whatever;"
`;

exports[`insertAfterImports no imports 1`] = `
"
INSERT

foo bar;
baz;"
`;

exports[`insertAfterImports single-line imports 1`] = `
"import 'foo';
import { bar } from 'baz';
INSERT

import 'foo';
import { bar } from 'baz';
whatever;"
`;
